units/en/unit1/2.md

```diff
@@ -0,0 +1,755 @@
+    {"role": "assistant", "content": '{\n  "question": "What\'s the capital of France?",\n  "answer": "'}
+]
+
+# The model will continue with just the answer, maintaining JSON structure
+```
+
+**2. Code Completion:**
+```python
+# Guide the model to complete code
+messages = [
+    {"role": "user", "content": "Write a Python function to calculate factorial"},
+    {"role": "assistant", "content": "def factorial(n):\n    if n == 0:\n        return 1\n    else:\n        return n * "}
+]
+
+# Model will complete the recursive call
+```
+
+**3. Step-by-Step Reasoning:**
+```python
+# Guide the model through structured thinking
+messages = [
+    {"role": "user", "content": "Solve: 2x + 5 = 13"},
+    {"role": "assistant", "content": "Let me solve this step by step:\n\nStep 1: "}
+]
+
+# Model will continue with the first step
+```
+
+### Important Notes
+
+- You cannot use `add_generation_prompt=True` and `continue_final_message=True` together
+- The final message must have the "assistant" role when using `continue_final_message=True`
+- This feature removes end-of-sequence tokens from the final message
+
+## Working with Reasoning Mode
+
+SmolLM3's dual-mode reasoning can be controlled through special formatting:
+
+### Standard vs Thinking Mode
+
+```python
+# Standard mode - direct answer
+standard_messages = [
+    {"role": "user", "content": "What is 15 × 24?"},
+    {"role": "assistant", "content": "15 × 24 = 360"}
+]
+
+# Thinking mode - show reasoning process
+thinking_messages = [
+    {"role": "user", "content": "What is 15 × 24?"},
+    {"role": "assistant", "content": "<|thinking|>\nI need to multiply 15 by 24. Let me break this down:\n15 × 24 = 15 × (20 + 4) = (15 × 20) + (15 × 4) = 300 + 60 = 360\n</|thinking|>\n\n15 × 24 = 360"}
+]
+
+# Apply templates
+standard_formatted = tokenizer.apply_chat_template(standard_messages, tokenize=False)
+thinking_formatted = tokenizer.apply_chat_template(thinking_messages, tokenize=False)
+
+print("Standard mode:")
+print(standard_formatted)
+print("\nThinking mode:")
+print(thinking_formatted)
+```
+
+### Training with Thinking Mode
+
+When preparing datasets with thinking mode, you can control whether to include the reasoning:
+
+```python
+def create_thinking_example(question, answer, reasoning=None):
+    """Create a training example with optional thinking"""
+    if reasoning:
+        assistant_content = f"<|thinking|>\n{reasoning}\n</|thinking|>\n\n{answer}"
+    else:
+        assistant_content = answer
+    
+    return [
+        {"role": "user", "content": question},
+        {"role": "assistant", "content": assistant_content}
+    ]
+
+# Example usage
+math_example = create_thinking_example(
+    question="What is the derivative of x²?",
+    answer="The derivative of x² is 2x",
+    reasoning="Using the power rule: d/dx(x^n) = n·x^(n-1)\nFor x²: n=2, so d/dx(x²) = 2·x^(2-1) = 2x"
+)
+```
+
+## Pipeline Usage: Automated Chat Processing
```

Comment:
brainstorming: I wonder if it's easier to start with a simpler `pipeline` style first and then move to tokenizer + generate way